gui/RespiraConnectedDevice.cpp

```diff
@@ -0,0 +1,83 @@
+#include "../common/third_party/nanopb/pb_common.h"
+#include "../common/third_party/nanopb/pb_decode.h"
+#include "../common/third_party/nanopb/pb_encode.h"
+#include "connected_device.h"
+#include <QSerialPort>
+
+class RespiraConnectedDevice : public ConnectedDevice {
+
+  RespiraConnectedDevice(QString portName, uint32_t write_timeout_ms,
+                         uint32_t interframe_timeout_ms)
+      : m_serial(new QSerialPort()), m_writeTimeout_ms(write_timeout_ms),
+        m_interframeTimeout_ms(interframe_timeout_ms) {
+    m_serial->setPortName(portName);
+    m_serial->setBaudRate(QSerialPort::Baud115200);
+    m_serial->setDataBits(QSerialPort::Data8);
+    m_serial->setParity(QSerialPort::NoParity);
+    m_serial->setStopBits(QSerialPort::OneStop);
+    m_serial->setFlowControl(QSerialPort::NoFlowControl);
+    m_serial->open(QIODevice::ReadWrite);
+  }
+
+public:
+  void SendGuiStatus(const GuiStatus &gui_status) override {
+
+    uint8_t tx_buffer[GuiStatus_size];
+    pb_ostream_t stream = pb_ostream_from_buffer(tx_buffer, sizeof(tx_buffer));
+    if (!pb_encode(&stream, GuiStatus_fields, &gui_status)) {
+      // TODO: Serialization failure; log an error or raise an alert.
+      return;
+    }
+
+    m_serial->write((const char *)tx_buffer, stream.bytes_written);
+
+    if (!m_serial->waitForBytesWritten(m_writeTimeout_ms)) {
+      // TODO communication failure, port closed? Log an error and raise an
+      // alert
+      return;
+    }
+  }
+
+  void ReceiveControllerStatus(ControllerStatus *controller_status) override {
+    // Drop everything we have in buffer right now as we are not sure if this is
+    // a full frame or a part of it.
+    // We drop everything that comes in until we get at least half of interframe
+    // silence
+    if (m_serial->bytesAvailable() > 0) {
+      while (m_serial->waitForReadyRead(m_interframeTimeout_ms / 2)) {
+        m_serial->clear(QSerialPort::Input);
+      }
+    }
+
+    if (m_serial->waitForReadyRead(m_interframeTimeout_ms)) {
+      QByteArray responseData = m_serial->readAll();
```

Comment:
Can we preallocate this with `.reserve()`? I guess ControllerStatus might be small enough that it doesn't matter though.